lib/Sema/NameBinding.cpp

```diff
@@ -36,60 +37,539 @@
+
+  if (!options.contains(ImportFlags::PrivateImport))
+    return;
+
+  if (topLevelModule->arePrivateImportsEnabled())
+    return;
+
+  diagnoseInvalidAttr(DAK_PrivateImport, ctx.Diags,
+                      diag::module_not_compiled_for_private_import);
+  privateImportFileName = StringRef();
+}
+
+void UnboundImport::validateImplementationOnly(ASTContext &ctx) {
+  if (!options.contains(ImportFlags::ImplementationOnly) ||
+      !options.contains(ImportFlags::Exported))
+    return;
+
+  // Remove one flag to maintain the invariant.
+  options -= ImportFlags::ImplementationOnly;
+
+  diagnoseInvalidAttr(DAK_ImplementationOnly, ctx.Diags,
+                      diag::import_implementation_cannot_be_exported);
+}
+
+void UnboundImport::validateTestable(ModuleDecl *topLevelModule) {
+  assert(topLevelModule);
+  ASTContext &ctx = topLevelModule->getASTContext();
+
+  if (!options.contains(ImportFlags::Testable) ||
+      topLevelModule->isTestingEnabled() ||
+      topLevelModule->isNonSwiftModule() ||
+      !ctx.LangOpts.EnableTestableAttrRequiresTestableModule)
+    return;
+
+  diagnoseInvalidAttr(DAK_Testable, ctx.Diags, diag::module_not_testable);
+}
+
+void UnboundImport::validateResilience(NullablePtr<ModuleDecl> topLevelModule,
+                                       SourceFile &SF) {
+  if (options.contains(ImportFlags::ImplementationOnly))
+    return;
+
+  // Per getTopLevelModule(), we'll only get nullptr here for non-Swift modules,
+  // so these two really mean the same thing.
+  if (!topLevelModule || topLevelModule.get()->isNonSwiftModule())
+    return;
+
+  if (!SF.getParentModule()->isResilient() ||
+      topLevelModule.get()->isResilient())
+    return;
+
+  ASTContext &ctx = SF.getASTContext();
+  ctx.Diags.diagnose(modulePath.front().Loc,
+                     diag::module_not_compiled_with_library_evolution,
+                     topLevelModule.get()->getName(),
+                     SF.getParentModule()->getName());
+  // FIXME: Once @_implementationOnly is a real feature, we should have a fix-it
+  // to add it.
+}
+
+void UnboundImport::diagnoseInvalidAttr(DeclAttrKind attrKind,
+                                        DiagnosticEngine &diags,
+                                        Diag<Identifier> diagID) {
+  auto diag = diags.diagnose(modulePath.front().Loc, diagID,
+                             modulePath.front().Item);
+
+  if (!attrs) return;
+  auto *attr = attrs.get()->getAttribute(attrKind);
+  if (!attr) return;
+
+  diag.fixItRemove(attr->getRangeWithAt());
+  attr->setInvalid();
+}
+
+//===----------------------------------------------------------------------===//
+// MARK: Scoped imports
+//===----------------------------------------------------------------------===//
+
+BoundImport::BoundImport(UnboundImport unbound, ImportedModuleDesc desc,
+                         ModuleDecl *module, bool needsScopeValidation)
+  : unbound(unbound), desc(desc), module(module),
```

Comment:
Is it ok for the `ModuleDecl *` to be null (e.g. in the case of mixed-source frameworks?)? If not, it would be good to add an assertion here.

Reply:
I actually did this in response to one of Hamish's review comments.